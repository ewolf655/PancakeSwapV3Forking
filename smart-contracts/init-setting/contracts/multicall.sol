// SPDX-License-Identifier: MIT
pragma solidity =0.8.24;
pragma abicoder v2;

/// @notice A fork of MultiCall with gas limit
contract PancakeInterfaceMulticallV2 {
    struct Call {
        address target;
        uint256 gasLimit;
        bytes callData;
    }

    struct Result {
        bool success;
        uint256 gasUsed;
        bytes returnData;
    }

    function multicall(Call[] memory calls) public returns (uint256 blockNumber, Result[] memory returnData) {
        blockNumber = block.number;
        returnData = new Result[](calls.length);

        for (uint256 i = 0; i < calls.length; i++) {
            (address target, uint256 gasLimit, bytes memory callData) = (
                calls[i].target,
                calls[i].gasLimit,
                calls[i].callData
            );

            uint256 gasLeftBefore = gasleft();
            (bool success, bytes memory ret) = target.call{gas: gasLimit}(callData);
            uint256 gasUsed = gasLeftBefore - gasleft();
            returnData[i] = Result(success, gasUsed, ret);
        }
    }

    function multicallWithGasLimitation(
        Call[] memory calls,
        uint256 gasBuffer
    ) public returns (uint256 blockNumber, Result[] memory returnData, uint256 lastSuccessIndex) {
        blockNumber = block.number;
        returnData = new Result[](calls.length);

        for (uint256 i = 0; i < calls.length; i++) {
            (address target, uint256 gasLimit, bytes memory callData) = (
                calls[i].target,
                calls[i].gasLimit,
                calls[i].callData
            );

            uint256 gasLeftBefore = gasleft();
            (bool success, bytes memory ret) = target.call{gas: gasLimit}(callData);
            uint256 gasUsed = gasLeftBefore - gasleft();
            returnData[i] = Result(success, gasUsed, ret);

            if (gasleft() < gasBuffer) {
                return (blockNumber, returnData, i);
            }
        }

        return (blockNumber, returnData, calls.length - 1);
    }

    function gaslimit() external view returns (uint256) {
        return block.gaslimit;
    }

    function gasLeft() external view returns (uint256) {
        return gasleft();
    }
}