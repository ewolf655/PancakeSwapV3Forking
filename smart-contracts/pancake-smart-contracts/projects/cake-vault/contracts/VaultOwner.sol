// SPDX-License-Identifier: MIT
pragma solidity 0.6.12;

import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
import "./CakeVault.sol";

contract VaultOwner is Ownable {
    using SafeERC20 for IERC20;

    CakeVault public immutable cakeVault;

    /**
     * @notice Constructor
     * @param _cakeVaultAddress: CakeVault contract address
     */
    constructor(address _cakeVaultAddress) public {
        cakeVault = CakeVault(_cakeVaultAddress);
    }

    /**
     * @notice Sets admin address to this address
     * @dev Only callable by the contract owner.
     * It makes the admin == owner.
     */
    function setAdmin() external onlyOwner {
        cakeVault.setAdmin(address(this));
    }

    /**
     * @notice Sets treasury address
     * @dev Only callable by the contract owner.
     */
    function setTreasury(address _treasury) external onlyOwner {
        cakeVault.setTreasury(_treasury);
    }

    /**
     * @notice Sets performance fee
     * @dev Only callable by the contract owner.
     */
    function setPerformanceFee(uint256 _performanceFee) external onlyOwner {
        cakeVault.setPerformanceFee(_performanceFee);
    }

    /**
     * @notice Sets call fee
     * @dev Only callable by the contract owner.
     */
    function setCallFee(uint256 _callFee) external onlyOwner {
        cakeVault.setCallFee(_callFee);
    }

    /**
     * @notice Sets withdraw fee
     * @dev Only callable by the contract owner.
     */
    function setWithdrawFee(uint256 _withdrawFee) external onlyOwner {
        cakeVault.setWithdrawFee(_withdrawFee);
    }

    /**
     * @notice Sets withdraw fee period
     * @dev Only callable by the contract owner.
     */
    function setWithdrawFeePeriod(uint256 _withdrawFeePeriod) external onlyOwner {
        cakeVault.setWithdrawFeePeriod(_withdrawFeePeriod);
    }

    /**
     * @notice Withdraw unexpected tokens sent to the Cake Vault
     */
    function inCaseTokensGetStuck(address _token) external onlyOwner {
        cakeVault.inCaseTokensGetStuck(_token);
        uint256 amount = IERC20(_token).balanceOf(address(this));
        IERC20(_token).safeTransfer(msg.sender, amount);
    }

    /**
     * @notice Triggers stopped state
     * @dev Only possible when contract not paused.
     */
    function pause() external onlyOwner {
        cakeVault.pause();
    }

    /**
     * @notice Returns to normal state
     * @dev Only possible when contract is paused.
     */
    function unpause() external onlyOwner {
        cakeVault.unpause();
    }
}
