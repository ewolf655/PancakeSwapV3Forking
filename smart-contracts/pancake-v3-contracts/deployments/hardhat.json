{
  "MasterChefV3": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  "SmartRouter": "0x5579c645590c58E6421341394f82B5b38080cDEA",
  "SmartRouterHelper": "0xA2b9C533b5333c1fEc593DA3Dbe1229627490622",
  "MixedRouteQuoterV1": "0x0a7DE23cF80aBA83Ca5d4a429D80E1224b3dCC91",
  "QuoterV2": "0x2279B7A0a67DB372996a5FaB50D91eAA73d2eBe6",
  "TokenValidator": "0x023108300557958e6eEf0DCc69e2D3804A6F13a1",
  "PancakeV3Factory": "0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512",
  "PancakeV3PoolDeployer": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  "PancakeV3FactoryOwner": "0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9",
  "SwapRouter": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  "V3Migrator": "0x0165878A594ca255338adfa4d48449f69242Eb8F",
  "TickLens": "0xa513E6E4b8f2a923D98304ec87F64353C4D5C853",
  "ProxyAdmin": "0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0",
  "TransparentUpgradeableProxy": "0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9",
  "NonfungibleTokenPositionDescriptor": "0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512",
  "NonfungiblePositionManager": "0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9",
  "PancakeInterfaceMulticall": "0x5FC8d32690cc91D4c39d9d3abcBD16989F875707",
  "PancakeV3LmPoolDeployer": "0x5FbDB2315678afecb367f032d93F642f64180aa3"
}
