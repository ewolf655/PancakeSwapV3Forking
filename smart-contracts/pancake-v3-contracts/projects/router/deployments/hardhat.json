{
  "SmartRouter": "0x5579c645590c58E6421341394f82B5b38080cDEA",
  "SmartRouterHelper": "0xA2b9C533b5333c1fEc593DA3Dbe1229627490622",
  "MixedRouteQuoterV1": "0x0a7DE23cF80aBA83Ca5d4a429D80E1224b3dCC91",
  "QuoterV2": "0x9163fe1E8448c237d5e7CEbF428EbA3015841869",
  "TokenValidator": "0x023108300557958e6eEf0DCc69e2D3804A6F13a1"
}