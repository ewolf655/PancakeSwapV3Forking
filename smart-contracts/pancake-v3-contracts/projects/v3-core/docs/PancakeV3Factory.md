# Solidity API

## PancakeV3Factory

Deploys PancakeSwap V3 pools and manages ownership and control over pool protocol fees

### owner

```solidity
address owner
```

Returns the current owner of the factory

_Can be changed by the current owner via setOwner_

#### Return Values

| Name | Type | Description |
| ---- | ---- | ----------- |

### poolDeployer

```solidity
address poolDeployer
```

Returns the current pool deployer

### feeAmountTickSpacing

```solidity
mapping(uint24 => int24) feeAmountTickSpacing
```

Returns the tick spacing for a given fee amount, if enabled, or 0 if not enabled

_A fee amount can never be removed, so this value should be hard coded or cached in the calling context_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |

#### Return Values

| Name | Type | Description |
| ---- | ---- | ----------- |

### getPool

```solidity
mapping(address => mapping(address => mapping(uint24 => address))) getPool
```

Returns the pool address for a given pair of tokens and a fee, or address 0 if it does not exist

_tokenA and tokenB may be passed in either token0/token1 or token1/token0 order_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |

#### Return Values

| Name | Type | Description |
| ---- | ---- | ----------- |

### feeAmountTickSpacingExtraInfo

```solidity
mapping(uint24 => struct IPancakeV3Factory.TickSpacingExtraInfo) feeAmountTickSpacingExtraInfo
```

Returns the tick spacing extra info

_A fee amount can never be removed, so this value should be hard coded or cached in the calling context_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |

#### Return Values

| Name | Type | Description |
| ---- | ---- | ----------- |

### constructor

```solidity
constructor(address _poolDeployer) public
```

### createPool

```solidity
function createPool(address tokenA, address tokenB, uint24 fee) external returns (address pool)
```

Creates a pool for the given two tokens and fee

_tokenA and tokenB may be passed in either order: token0/token1 or token1/token0. tickSpacing is retrieved
from the fee. The call will revert if the pool already exists, the fee is invalid, or the token arguments
are invalid._

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| tokenA | address | One of the two tokens in the desired pool |
| tokenB | address | The other of the two tokens in the desired pool |
| fee | uint24 | The desired fee for the pool |

#### Return Values

| Name | Type | Description |
| ---- | ---- | ----------- |
| pool | address | The address of the newly created pool |

### setOwner

```solidity
function setOwner(address _owner) external
```

Updates the owner of the factory

_Must be called by the current owner_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| _owner | address | The new owner of the factory |

### enableFeeAmount

```solidity
function enableFeeAmount(uint24 fee, int24 tickSpacing) public
```

Enables a fee amount with the given tickSpacing

_Fee amounts may never be removed once enabled_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| fee | uint24 | The fee amount to enable, denominated in hundredths of a bip (i.e. 1e-6) |
| tickSpacing | int24 | The spacing between ticks to be enforced for all pools created with the given fee amount |

### setWhiteListAddress

```solidity
function setWhiteListAddress(address user, bool verified) public
```

Set an address into white list

_Address can be updated by owner with boolean value false_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| user | address | The user address that add into white list |
| verified | bool |  |

### setFeeAmountExtraInfo

```solidity
function setFeeAmountExtraInfo(uint24 fee, bool whitelistRequested, bool enabled) public
```

Set a fee amount extra info

_Fee amounts can be updated by owner with extra info_

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| fee | uint24 |  |
| whitelistRequested | bool | The flag whether should be created by owner only |
| enabled | bool | The flag is the fee is enabled or not |

