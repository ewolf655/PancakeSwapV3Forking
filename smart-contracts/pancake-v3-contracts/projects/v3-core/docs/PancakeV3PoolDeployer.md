# Solidity API

## PancakeV3PoolDeployer

### Parameters

```solidity
struct Parameters {
  address factory;
  address token0;
  address token1;
  uint24 fee;
  int24 tickSpacing;
}
```

### parameters

```solidity
struct PancakeV3PoolDeployer.Parameters parameters
```

Get the parameters to be used in constructing the pool, set transiently during pool creation.

_Called by the pool constructor to fetch the parameters of the pool
Returns factory The factory address
Returns token0 The first token of the pool by address sort order
Returns token1 The second token of the pool by address sort order
Returns fee The fee collected upon every swap in the pool, denominated in hundredths of a bip
Returns tickSpacing The minimum number of ticks between initialized ticks_

### factoryAddress

```solidity
address factoryAddress
```

### SetFactoryAddress

```solidity
event SetFactoryAddress(address factory)
```

Emitted when factory address is set

### onlyFactory

```solidity
modifier onlyFactory()
```

### setFactoryAddress

```solidity
function setFactoryAddress(address _factoryAddress) external
```

### deploy

```solidity
function deploy(address factory, address token0, address token1, uint24 fee, int24 tickSpacing) external returns (address pool)
```

_Deploys a pool with the given parameters by transiently setting the parameters storage slot and then
clearing it after deploying the pool._

#### Parameters

| Name | Type | Description |
| ---- | ---- | ----------- |
| factory | address | The contract address of the PancakeSwap V3 factory |
| token0 | address | The first token of the pool by address sort order |
| token1 | address | The second token of the pool by address sort order |
| fee | uint24 | The fee collected upon every swap in the pool, denominated in hundredths of a bip |
| tickSpacing | int24 | The spacing between usable ticks |

