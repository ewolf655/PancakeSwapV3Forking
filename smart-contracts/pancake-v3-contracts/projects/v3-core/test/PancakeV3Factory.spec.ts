import { Wallet } from 'ethers'
import { ethers, waffle } from 'hardhat'
import { PancakeV3Factory } from '../typechain-types/contracts/PancakeV3Factory'
import { PancakeV3PoolDeployer } from '../typechain-types/contracts/PancakeV3PoolDeployer'
import { expect } from './shared/expect'
import snapshotGasCost from './shared/snapshotGasCost'

import { FeeAmount, getCreate2Address, TICK_SPACINGS } from './shared/utilities'

const { constants } = ethers

const TEST_ADDRESSES: [string, string] = [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000',
]

const createFixtureLoader = waffle.createFixtureLoader

describe('PancakeV3Factory', () => {
  let wallet: Wallet, other: Wallet

  let deployer: PancakeV3PoolDeployer
  let factory: PancakeV3Factory
  let poolBytecode: string
  const fixture = async () => {
    const deployerFactory = await ethers.getContractFactory('PancakeV3PoolDeployer')
    const factoryFactory = await ethers.getContractFactory('PancakeV3Factory')
    const deployer_ = (await deployerFactory.deploy()) as PancakeV3PoolDeployer
    const factory_ = (await factoryFactory.deploy(deployer_.address)) as PancakeV3Factory

    await deployer_.setFactoryAddress(factory_.address)
    return [factory_, deployer_] as [PancakeV3Factory, PancakeV3PoolDeployer]
  }

  let loadFixture: ReturnType<typeof createFixtureLoader>
  before('create fixture loader', async () => {
    ;[wallet, other] = await (ethers as any).getSigners()

    loadFixture = createFixtureLoader([wallet, other])
  })

  before('load pool bytecode', async () => {
    poolBytecode = (await ethers.getContractFactory('PancakeV3Pool')).bytecode
  })

  beforeEach('deploy factory', async () => {
    ;[factory, deployer] = await loadFixture(fixture)
  })

  it('owner is deployer', async () => {
    expect(await factory.owner()).to.eq(wallet.address)
  })

  it('factory bytecode size', async () => {
    expect(((await waffle.provider.getCode(factory.address)).length - 2) / 2).to.matchSnapshot()
  })

  it('pool bytecode size', async () => {
    await factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[1], FeeAmount.MEDIUM)
    const poolAddress = getCreate2Address(deployer.address, TEST_ADDRESSES, FeeAmount.MEDIUM, poolBytecode)
    expect(((await waffle.provider.getCode(poolAddress)).length - 2) / 2).to.matchSnapshot()
  })

  it('initial enabled fee amounts', async () => {
    expect(await factory.feeAmountTickSpacing(FeeAmount.LOW)).to.eq(TICK_SPACINGS[FeeAmount.LOW])
    expect(await factory.feeAmountTickSpacing(FeeAmount.MEDIUM)).to.eq(TICK_SPACINGS[FeeAmount.MEDIUM])
    expect(await factory.feeAmountTickSpacing(FeeAmount.HIGH)).to.eq(TICK_SPACINGS[FeeAmount.HIGH])
  })

  async function createAndCheckPool(
    tokens: [string, string],
    feeAmount: FeeAmount,
    tickSpacing: number = TICK_SPACINGS[feeAmount]
  ) {
    const create2Address = getCreate2Address(deployer.address, tokens, feeAmount, poolBytecode)
    const create = factory.createPool(tokens[0], tokens[1], feeAmount)

    await expect(create)
      .to.emit(factory, 'PoolCreated')
      .withArgs(TEST_ADDRESSES[0], TEST_ADDRESSES[1], feeAmount, tickSpacing, create2Address)

    await expect(factory.createPool(tokens[0], tokens[1], feeAmount)).to.be.reverted
    await expect(factory.createPool(tokens[1], tokens[0], feeAmount)).to.be.reverted
    expect(await factory.getPool(tokens[0], tokens[1], feeAmount), 'getPool in order').to.eq(create2Address)
    expect(await factory.getPool(tokens[1], tokens[0], feeAmount), 'getPool in reverse').to.eq(create2Address)

    const poolContractFactory = await ethers.getContractFactory('PancakeV3Pool')
    const pool = poolContractFactory.attach(create2Address)
    expect(await pool.factory(), 'pool factory address').to.eq(factory.address)
    expect(await pool.token0(), 'pool token0').to.eq(TEST_ADDRESSES[0])
    expect(await pool.token1(), 'pool token1').to.eq(TEST_ADDRESSES[1])
    expect(await pool.fee(), 'pool fee').to.eq(feeAmount)
    expect(await pool.tickSpacing(), 'pool tick spacing').to.eq(tickSpacing)
  }

  describe('#createPool', () => {
    it('succeeds for low fee pool', async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.LOW)
    })

    it('succeeds for medium fee pool', async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.MEDIUM)
    })
    it('succeeds for high fee pool', async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.HIGH)
    })

    it('succeeds if tokens are passed in reverse', async () => {
      await createAndCheckPool([TEST_ADDRESSES[1], TEST_ADDRESSES[0]], FeeAmount.MEDIUM)
    })

    it('fails if token a == token b', async () => {
      await expect(factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[0], FeeAmount.LOW)).to.be.reverted
    })

    it('fails if token a is 0 or token b is 0', async () => {
      await expect(factory.createPool(TEST_ADDRESSES[0], constants.AddressZero, FeeAmount.LOW)).to.be.reverted
      await expect(factory.createPool(constants.AddressZero, TEST_ADDRESSES[0], FeeAmount.LOW)).to.be.reverted
      await expect(factory.createPool(constants.AddressZero, constants.AddressZero, FeeAmount.LOW)).to.be.revertedWith(
        ''
      )
    })

    it('fails if fee amount is not enabled', async () => {
      await expect(factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[1], 250)).to.be.reverted
    })

    it('gas', async () => {
      await snapshotGasCost(factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[1], FeeAmount.MEDIUM))
    })
  })

  describe('#setOwner', () => {
    it('fails if caller is not owner', async () => {
      await expect(factory.connect(other).setOwner(wallet.address)).to.be.reverted
    })

    it('updates owner', async () => {
      await factory.setOwner(other.address)
      expect(await factory.owner()).to.eq(other.address)
    })

    it('emits event', async () => {
      await expect(factory.setOwner(other.address))
        .to.emit(factory, 'OwnerChanged')
        .withArgs(wallet.address, other.address)
    })

    it('cannot be called by original owner', async () => {
      await factory.setOwner(other.address)
      await expect(factory.setOwner(wallet.address)).to.be.reverted
    })
  })

  describe('#enableFeeAmount', () => {
    it('fails if caller is not owner', async () => {
      await expect(factory.connect(other).enableFeeAmount(200, 2)).to.be.reverted
    })
    it('fails if fee is too great', async () => {
      await expect(factory.enableFeeAmount(1000000, 10)).to.be.reverted
    })
    it('fails if tick spacing is too small', async () => {
      await expect(factory.enableFeeAmount(500, 0)).to.be.reverted
    })
    it('fails if tick spacing is too large', async () => {
      await expect(factory.enableFeeAmount(500, 16834)).to.be.reverted
    })
    it('fails if already initialized', async () => {
      await expect(factory.enableFeeAmount(100, 10)).to.be.reverted
    })
    it('sets the fee amount in the mapping', async () => {
      await factory.enableFeeAmount(200, 5)
      expect(await factory.feeAmountTickSpacing(200)).to.eq(5)
    })
    it('emits an event', async () => {
      await expect(factory.enableFeeAmount(200, 5)).to.emit(factory, 'FeeAmountEnabled').withArgs(200, 5)
    })
    it('enables pool creation', async () => {
      await factory.enableFeeAmount(250, 15)
      await createAndCheckPool([TEST_ADDRESSES[0], TEST_ADDRESSES[1]], 250, 15)
    })

    it('fails if enables pool creation without whitelist', async () => {
      await factory.enableFeeAmount(350, 15)
      await factory.setFeeAmountExtraInfo(350, true, true)
      await expect(createAndCheckPool([TEST_ADDRESSES[0], TEST_ADDRESSES[1]], 350, 15)).to.be.reverted
    })

    it('fails if enables pool creation without enable', async () => {
      await factory.enableFeeAmount(450, 15)
      await factory.setFeeAmountExtraInfo(450, true, false)
      await factory.setWhiteListAddress(wallet.address, true)
      await expect(createAndCheckPool([TEST_ADDRESSES[0], TEST_ADDRESSES[1]], 450, 15)).to.be.reverted
    })

    it('enables pool creation with enable and whitelist', async () => {
      await factory.enableFeeAmount(550, 15)
      await factory.setFeeAmountExtraInfo(550, true, true)
      await factory.setWhiteListAddress(wallet.address, true)
      createAndCheckPool([TEST_ADDRESSES[0], TEST_ADDRESSES[1]], 550, 15)
    })
  })
})
