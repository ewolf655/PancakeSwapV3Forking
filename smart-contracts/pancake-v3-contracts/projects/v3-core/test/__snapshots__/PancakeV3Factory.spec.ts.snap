// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PancakeV3Factory #createPool gas 1`] = `4729417`;

exports[`PancakeV3Factory factory bytecode size 1`] = `5151`;

exports[`PancakeV3Factory pool bytecode size 1`] = `22962`;
